source/solvers/tracer.cc

```diff
@@ -530,6 +553,179 @@ Tracer<dim>::calculate_tracer_statistics()
+    this->simulation_parameters.physical_properties_manager;
+  std::map<field, std::vector<double>> fields;
+
+  const auto diffusivity_model = properties_manager.get_tracer_diffusivity();
+
+  std::vector<double>     tracer_diffusivity(n_q_points_face);
+  std::vector<Point<dim>> face_quadrature_points;
+
+  std::vector<double> tracer_flow_rate_vector(
+    this->simulation_parameters.boundary_conditions.size, 0);
+
+  // Get vector of all boundary conditions
+  const auto boundary_conditions_ids =
+    this->simulation_parameters.boundary_conditions.id;
+
+  for (const auto &cell : this->dof_handler.active_cell_iterators())
+    {
+      if (cell->is_locally_owned() && cell->at_boundary())
+        {
+          for (const auto face : cell->face_indices())
+            {
+              if (cell->face(face)->at_boundary())
+                {
+                  const auto boundary_id =
+                    std::find(begin(boundary_conditions_ids),
+                              end(boundary_conditions_ids),
+                              cell->face(face)->boundary_id());
+
+
+                  unsigned int vector_index =
+                    boundary_id - boundary_conditions_ids.begin();
+
+                  // Gather tracer information
+                  fe_face_values_tracer.reinit(cell, face);
+                  fe_face_values_tracer.get_function_values(
+                    this->present_solution, tracer_values);
+                  fe_face_values_tracer.get_function_gradients(
+                    this->present_solution, tracer_gradient);
+
+                  // We update the fields required by the diffusivity
+                  // model
+                  fields.clear();
+                  if (diffusivity_model->depends_on(field::levelset))
+                    {
+                      std::vector<double> levelset_values(n_q_points_face);
+                      fields.insert(
+                        std::pair<field, std::vector<double>>(field::levelset,
+                                                              n_q_points_face));
+                      face_quadrature_points =
+                        fe_face_values_tracer.get_quadrature_points();
+                      immersed_solid_signed_distance_function =
+                        this->multiphysics
+                          ->get_immersed_solid_signed_distance_function();
+                      this->immersed_solid_signed_distance_function->value_list(
+                        face_quadrature_points, levelset_values);
+                      set_field_vector(field::levelset,
+                                       levelset_values,
+                                       fields);
+                    }
+
+                  diffusivity_model->vector_value(fields, tracer_diffusivity);
+
+                  // Get fluid dynamics active cell iterator
+                  typename DoFHandler<dim>::active_cell_iterator cell_fd(
+                    &(*(this->triangulation)),
```

Comment:
I'm not a pro in c++ but if feel this is redundant? Doesn't `this -> triangulation` already return the address?

Reply:
It's to get the more general type Triangulation (for the function) instead of the parallel::DistributedTriangulationBase that we have

Reply:
Interesting!

Reply:
this->triangulation returns a std::shared_ptr. If you want to have the raw pointer you can do a &(*) trick to obtain the raw pointer behind :)